### DISCUSSION TITLE
Constraint and onValidate Differences

### DISCUSSION THREAD

**Discussion by jtan80813** [edmundhung/conform#687] [Q&A]
In your example here: https://conform.guide/tutorial

What's the difference between `constraint` and `onValidate`,
Do we still need `constraint` cause we already have `onValidate` that would validate our schema on client side.

```
  const [form, fields] = useForm({
    lastResult,
    constraint: getZodConstraint(schema),
    shouldValidate: 'onBlur',
    shouldRevalidate: 'onInput',
    onValidate({ formData }) {
      return parseWithZod(formData, { schema });
    },
  });
```

**Comment by edmundhung** [comment]
The constraints are mostly about accessibility as you are validating with zod. As it is usually coupled with your schema, conform offers an helper that let you infer all the validation attributes only.
